oracle/src/poseidon_3.rs

```diff
@@ -0,0 +1,174 @@
+/*****************************************************
+This file implements Poseidon Hash Function primitive
+with parameters rounds=54, width=3, rate=2, alpha=7
+******************************************************/
+
+use algebra::Field;
+use super::poseidon::{ArithmeticSpongeParams, Sponge, SpongeState,
+                      SpongeConstants};
+
+#[derive(Clone)]
+pub struct PlonkSpongeConstants {
+}
+
+impl SpongeConstants for PlonkSpongeConstants {
+    const ROUNDS_FULL: usize = 54;
+    const ROUNDS_PARTIAL: usize = 0;
+    const HALF_ROUNDS_FULL: usize = 0;
+    const SPONGE_CAPACITY: usize = 1;
+    const SPONGE_WIDTH: usize = 3;
+    const SPONGE_RATE: usize = 2;
+    const SPONGE_BOX: usize = 7;
+    const FULL_MDS: bool = true;
+}
+
+pub fn sbox<F : Field, SC: SpongeConstants>(x: F) -> F {
+    x.pow([SC::SPONGE_BOX as u64])
+}
+
+#[derive(Clone)]
```

Comment:
Can you share the code with the other poseidon implementations? Shouldn't it just be a matter of specifying different constants rather than repeating the implementation?

Reply:
Sure, no problem.  (I only cloned it this way because I saw that's how it was done between poseidon 3-wire and poseidon 5-wire.)

Just to confirm, you want it refactored a bit so they can all share the same implementation, but just use different parameters and different permutation functions?